logicalplan/coalesce.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/thanos-io/promql-engine/query"
+)
+
+type CoalesceOptimizer struct{}
+
+func (c CoalesceOptimizer) Optimize(expr Node, opts *query.Options) (Node, annotations.Annotations) {
+	numShards := opts.NumShards()
+
+	TraverseBottomUp(nil, &expr, func(parent, e *Node) bool {
+		switch t := (*e).(type) {
+		case *VectorSelector:
+			if parent != nil {
+				// we coalesce matrix selectors in a different branch
+				if _, ok := (*parent).(*MatrixSelector); ok {
+					return false
+				}
+			}
+			exprs := make([]Node, numShards)
+			for i := range numShards {
+				vs := t.Clone().(*VectorSelector)
+				vs.Shard = i
+				vs.NumShards = numShards
+				exprs[i] = vs
+			}
+			*e = &Coalesce{Exprs: exprs}
+		case *MatrixSelector:
```

Comment:
Do we plan to handle matrix selector without function?

Reply:
Plain matrix selector is unsupported by Thanos engine.